doc/filters.md

```diff
@@ -0,0 +1,188 @@
+# CKAN Query Filters spec
+
+The goal of this specification is to provide a common syntax for defining filters
+that need to be applied to queries performed. It is not tied to a particular API
+endpoint or search backend, and it aims to be simple and extensible.
+
+## Overview
+
+Filters are always provided as a dictionary or as a list of dictionaries.
+
+When provided as a dictionary, the top level keys can either be:
+
+* A field name
+* A top level operator: one of `$or` or `$and`
+
+For example:
+
+```
+"filters": {
+    "field1": "value",
+    "$or": [
+        {"field2": ["value1", "value2"]},
+        {"field3": {"gte": 9000, "lt": 1000}}
+    ]
+}
+```
+
+When provided as a list of dictionaries, these are combined with the OR operator. For instance:
+
+```
+"filters": [
+    {"field1": "value1"},
+    {"field2": "value2"},
+]
+```
+
+is interpreted as:
+
+```
+"filters": {
+  "$or": [
+    {"field1": "value1"},
+    {"field2": "value2"},
+  ]
+}
+```
+
+## Individual filters
+
+Individual filters have the form:
+
+```
+  {
+    "<field_name>": {"<operator1>" : <value1>, "<operator2>" : <value2>...}
+  }
+
+```
+
+Note that fields starting with the character `$` must be escaped with another `$` character:
+
+```
+filters: {
+    "$$some_field": "some_value"
+}
+```
+
+If the operator is omitted, the `eq` (for equality) operator is assumed:
+
+```
+  {
+    "field1": "value1"
+  }
+
+```
+
+is equivalent to:
+
+```
+  {
+    "field1": {"eq" : "value1"}
+  }
+
+```
+
+If filters contain a list of values, the `in` operator is assumed, e.g.:
+
+```
+  {
+    "field1": ["value1", "value2", "value3"]
+  }
+
+```
+
+is interpreted as:
+
+```
+  {
+    "field1": {"in": ["value1", "value2", "value3"]}
+  }
+
+```
+
+## Built-in operators
+
+The following operators are supported by default:
+
+| Notation | Operation                            |
+| -------- | ------------------------------------ |
+| `eq`     | Equal                                |
+| `gt`     | Greater than                         |
+| `gte`    | Greater than or equal                |
+| `lt`     | Lower than                           |
+| `lte`    | Lower than or equal                  |
+| `in`     | Matches one of the provided values   |
+
+TODO: Additional operators can be added by plugins
+
+## Top level operators
+
+The following default top operators can be used to combine filters:
+
+| Notation | Operation                                  |
+| -------- | ------------------------------------------ |
+| `$or`    | logical OR, any of the filters can match   |
+| `$and`   | logical AND, all of the filters must match |
+
+
+Top level operators must always contain lists of at least two filters, i.e. all of the 
+following are incorrect:
+
+```
+{
+  "$or": {"field1": "value1"},
+}
+
+{
+  "$and": [
+    {"field1": "value1"}
+  ]
+}
+
+{
+  "$and": ["value1", "value2", "value3"]
+}
+
```

Comment:
````suggestion

When a dict is given as the value of an `$or` filter operation each key is treated as a separate part of the OR operation:

```
"filters": {
  "$or": {
    "color": "red",
    "size": 10
  }
}
```

is equivalent to:
```
"filters": {
  "$or": [
    {"color": "red"},
    {"size": 10}
  ]
}
```

This rule does **not** extend to values within a list, e.g.
```
"filters": {
  "$or": [
    {"color": "yellow", "size": 9},
  ]
}
```

is equivalent to:
```
"filters": {
  "color": "yellow", "size": 9
}
```

````

Reply:
@amercader what do you think about this one? On the one hand it allows users to skip a bunch of extra dicts in their `$or` operations, but on the other hand someone might eventually try to `"$or": {"score": {"gt": 90}, "score": {"lt": 20}}` and our code will **silently ignore** one of the conditions.

Reply:
I think we need to draw a line somewhere and avoid complicating the spec and the implementations, specially if it can lead to edge cases like the one you mention. I quite like enforcing lists of dicts for the filter operators. Let's test ride this and we can always extend it later.